source/http/HttpConnection.cpp

```diff
@@ -81,6 +81,7 @@ namespace Aws
                 (void)connection;
                 /* now that we're shutting down, we can release the internal ref count. */
                 auto *callbackData = static_cast<ConnectionCallbackData *>(user_data);
+                callbackData->connection->m_good = false;
```

Comment:
We could add `aws_http_connection_good()` or something like that, which would get you *slightly* more accurate answers. (at the cost of a lock/check/unlock)

Since m_good isn't set false until connection_shutdown(), there's a chance of someone's stream failing, which triggers the connection to close internally, which calls stream_complete() for all the pending streams before connection_shutdown() is invoked. So the user's stream will fail, and they might give the connection back to the ConnectionMgr, but Connection->m_good is still true so ConnectionMgr vends the connection to another user, who can't use it because it's mid-shutdown.

Reply:
we'll come back to this added a TODO